### PR TITLE
Add preload function for fast execution when `predict` is called

### PR DESCRIPTION
**Pull request by ph1ash**
#### Current issue
When calling `predict`, it takes a few seconds to build the predictor & run `setup_model`. When running in a pipeline, it's ideal to have the everything loaded and ready to go when a request comes in.

#### Proposed solution
Add a `preload` function which simply creates the predictor without having to call `predict` first and wait.

I have read the CLA Document and I sign the CLA

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Added a new `preload` method for initializing the model for inference.

### 📊 Key Changes
- 🛠️ A new `preload` method has been introduced in the `engine/model.py`.
- 🔧 The `preload` method accepts additional keyword arguments (`**kwargs`) for configuration.
- 📈 The `preload` method sets a default confidence threshold (`conf`) of 0.25 unless overridden.

### 🎯 Purpose & Impact
- 🚀 This PR aims to enhance model performance by allowing users to preload the model before conducting inference, which could lead to faster response times when performing predictions.
- 🌐 This change allows greater flexibility and customization of the prediction process by letting users specify their configuration parameters upfront, potentially improving user experience as they won’t need to set up the model every time predictions are made.

### CODE REVIEW

**Comment by ph1ash** [comment]
recheck

**Comment by ph1ash** [comment]
recheck

File: ultralytics/yolo/engine/model.py [hunk 1]
```diff
@@ -138,6 +138,26 @@ def fuse(self):
...
+            None
+        """
+        overrides = self.overrides.copy()
+        overrides["conf"] = 0.25
+        overrides.update(kwargs)
+        overrides["mode"] = "predict"
+        overrides["save"] = kwargs.get("save", False)  # not save files by default
+        if not self.predictor:
+            self.predictor = self.PredictorClass(overrides=overrides)
+            self.predictor.setup_model(model=self.model)
```
Comment (akx) [inline ultralytics/yolo/engine/model.py:159]: Shouldn't `.predict()` just call `.preload()` instead of this being copy-pasted from there?

**Review by akx — CHANGES_REQUESTED** [review]

**Comment by AyushExel** [comment]
I guess a simpler solution would be to just pass a random tensor to warmup the model before running inference.
```
model = ...
model(torch.Tensor) # preload
```
Does this work for your use case?

**Comment by glenn-jocher** [comment]
@ph1ash yes this is a good point. In python when the model is loaded we don't really know what it will be used for (i.e. training, validation, export, predict, etc), so predictor is not actually created until you try to run a prediction, but you can always run a dummy prediction to warm it up at Ayush mentioned, and then subsequent calls should be faster.

**Comment by ph1ash** [comment]
Thanks for the suggestions!! That didn't work for me actually, but perhaps I'm doing something incorrectly - see below:
```python
    def __init__(self):
        self.model_weights = os.getenv("MODEL_WEIGHTS") or "yolov8l.pt"
        ultralytics.checks()
        self.model = YOLO(self.model_weights)
        self.model(torch.Tensor)
```

Which causes the following error:
```
Ultralytics YOLOv8.0.13 🚀 Python-3.10.6 torch-1.13.1+cu117 CUDA:0 (NVIDIA GeForce RTX 3070, 8191MiB)
Setup complete ✅ (12 CPUs, 15.6 GB RAM, 127.4/1006.9 GB disk)
Ultralytics YOLOv8.0.13 🚀 Python-3.10.6 torch-1.13.1+cu117 CUDA:0 (NVIDIA GeForce RTX 3070, 8191MiB)
Fusing... 
YOLOv8l summary: 268 layers, 43668288 parameters, 0 gradients, 165.2 GFLOPs
Traceback (most recent call last):
  File "/home/ph1ash/dexlab/sandau/hyperion/inference-api/python/inference.py", line 59, in <module>
    inference_instance = Inference()
  File "/home/ph1ash/dexlab/sandau/hyperion/inference-api/python/inference.py", line 27, in __init__
    self.model(torch.Tensor)
  File "/home/ph1ash/dexlab/sandau/hyperion/inference-api/.venv/lib/python3.10/site-packages/ultralytics/yolo/engine/model.py", line 59, in __call__
    return self.predict(source, stream, verbose, **kwargs)
  File "/home/ph1ash/dexlab/sandau/hyperion/inference-api/.venv/lib/python3.10/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/home/ph1ash/dexlab/sandau/hyperion/inference-api/.venv/lib/python3.10/site-packages/ultralytics/yolo/engine/model.py", line 161, in predict
    return self.predictor(source=source, stream=stream, verbose=verbose)
  File "/home/ph1ash/dexlab/sandau/hyperion/inference-api/.venv/lib/python3.10/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/home/ph1ash/dexlab/sandau/hyperion/inference-api/.venv/lib/python3.10/site-packages/ultralytics/yolo/engine/predictor.py", line 158, in __call__
    return list(self.stream_inference(source, model, verbose))  # merge list of Result into one
  File "/home/ph1ash/dexlab/sandau/hyperion/inference-api/.venv/lib/python3.10/site-packages/ultralytics/yolo/engine/predictor.py", line 173, in stream_inference
    self.setup_source(source)
  File "/home/ph1ash/dexlab/sandau/hyperion/inference-api/.venv/lib/python3.10/site-packages/ultralytics/yolo/engine/predictor.py", line 133, in setup_source
    self.dataset = LoadPilAndNumpy(source,
  File "/home/ph1ash/dexlab/sandau/hyperion/inference-api/.venv/lib/python3.10/site-packages/ultralytics/yolo/data/dataloaders/stream_loaders.py", line 265, in __init__
    self.im0 = [self._single_check(im) for im in im0]
  File "/home/ph1ash/dexlab/sandau/hyperion/inference-api/.venv/lib/python3.10/site-packages/ultralytics/yolo/data/dataloaders/stream_loaders.py", line 265, in <listcomp>
    self.im0 = [self._single_check(im) for im in im0]
  File "/home/ph1ash/dexlab/sandau/hyperion/inference-api/.venv/lib/python3.10/site-packages/ultralytics/yolo/data/dataloaders/stream_loaders.py", line 276, in _single_check
    assert isinstance(im, (Image.Image, np.ndarray)), f"Expected PIL/np.ndarray image type, but got {type(im)}"
AssertionError: Expected PIL/np.ndarray image type, but got <class 'torch._C._TensorMeta'>
```

I believe this is because `model(torch.Tensor)` is eventually calling the `__call__` function of the `YOLO` class. Is there perhaps a different recommendation that would circumvent the need for the `preload` function?

**Comment by AyushExel** [comment]
> Thanks for the suggestions!! That didn't work for me actually, but perhaps I'm doing something incorrectly - see below:
> 
> ```python
>     def __init__(self):
>         self.model_weights = os.getenv("MODEL_WEIGHTS") or "yolov8l.pt"
>         ultralytics.checks()
>         self.model = YOLO(self.model_weights)
>         self.model(torch.Tensor)
> ```
> 
> Which causes the following error:
> 
> ```
> Ultralytics YOLOv8.0.13 🚀 Python-3.10.6 torch-1.13.1+cu117 CUDA:0 (NVIDIA GeForce RTX 3070, 8191MiB)
> Setup complete ✅ (12 CPUs, 15.6 GB RAM, 127.4/1006.9 GB disk)
> Ultralytics YOLOv8.0.13 🚀 Python-3.10.6 torch-1.13.1+cu117 CUDA:0 (NVIDIA GeForce RTX 3070, 8191MiB)
> Fusing... 
> YOLOv8l summary: 268 layers, 43668288 parameters, 0 gradients, 165.2 GFLOPs
> Traceback (most recent call last):
>   File "/home/ph1ash/dexlab/sandau/hyperion/inference-api/python/inference.py", line 59, in <module>
>     inference_instance = Inference()
>   File "/home/ph1ash/dexlab/sandau/hyperion/inference-api/python/inference.py", line 27, in __init__
>     self.model(torch.Tensor)
>   File "/home/ph1ash/dexlab/sandau/hyperion/inference-api/.venv/lib/python3.10/site-packages/ultralytics/yolo/engine/model.py", line 59, in __call__
>     return self.predict(source, stream, verbose, **kwargs)
>   File "/home/ph1ash/dexlab/sandau/hyperion/inference-api/.venv/lib/python3.10/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
>     return func(*args, **kwargs)
>   File "/home/ph1ash/dexlab/sandau/hyperion/inference-api/.venv/lib/python3.10/site-packages/ultralytics/yolo/engine/model.py", line 161, in predict
>     return self.predictor(source=source, stream=stream, verbose=verbose)
>   File "/home/ph1ash/dexlab/sandau/hyperion/inference-api/.venv/lib/python3.10/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
>     return func(*args, **kwargs)
>   File "/home/ph1ash/dexlab/sandau/hyperion/inference-api/.venv/lib/python3.10/site-packages/ultralytics/yolo/engine/predictor.py", line 158, in __call__
>     return list(self.stream_inference(source, model, verbose))  # merge list of Result into one
>   File "/home/ph1ash/dexlab/sandau/hyperion/inference-api/.venv/lib/python3.10/site-packages/ultralytics/yolo/engine/predictor.py", line 173, in stream_inference
>     self.setup_source(source)
>   File "/home/ph1ash/dexlab/sandau/hyperion/inference-api/.venv/lib/python3.10/site-packages/ultralytics/yolo/engine/predictor.py", line 133, in setup_source
>     self.dataset = LoadPilAndNumpy(source,
>   File "/home/ph1ash/dexlab/sandau/hyperion/inference-api/.venv/lib/python3.10/site-packages/ultralytics/yolo/data/dataloaders/stream_loaders.py", line 265, in __init__
>     self.im0 = [self._single_check(im) for im in im0]
>   File "/home/ph1ash/dexlab/sandau/hyperion/inference-api/.venv/lib/python3.10/site-packages/ultralytics/yolo/data/dataloaders/stream_loaders.py", line 265, in <listcomp>
>     self.im0 = [self._single_check(im) for im in im0]
>   File "/home/ph1ash/dexlab/sandau/hyperion/inference-api/.venv/lib/python3.10/site-packages/ultralytics/yolo/data/dataloaders/stream_loaders.py", line 276, in _single_check
>     assert isinstance(im, (Image.Image, np.ndarray)), f"Expected PIL/np.ndarray image type, but got {type(im)}"
> AssertionError: Expected PIL/np.ndarray image type, but got <class 'torch._C._TensorMeta'>
> ```
> 
> I believe this is because `model(torch.Tensor)` is eventually calling the `__call__` function of the `YOLO` class. Is there perhaps a different recommendation that would circumvent the need for the `preload` function?

Ohh that was just an exmple.. You'll need an actuall tensor. Try this:
```
    tens = torch.zeros(320, 640, 3)
    model.predict(tens)
```

**Comment by ph1ash** [comment]
When using the example above, I receive the following error:
 ```
 AssertionError: Expected PIL/np.ndarray image type, but got <class 'torch.Tensor'>
 ```
 
 I instead used:
 ```python
 tens = np.zeros((320, 640, 3), 'uint8')
 ```
 
 which works without issue and I'm able to preload the model for prediction. I'd be up for creating a small example mentioning this method if that'd help?

**Comment by AyushExel** [comment]
Oh yes, torch inference is not supported yet. The np input should do the trick!

**Pull request closed without merging**